CRM-21204 : Add progress bar for Contribution, membership and activity import

jitendrapurohit opened the pull request:
Overview
----------------------------------------
This PR adds progress bar for activity, contribution and membership imports.

Before
----------------------------------------
- Progress bar is only shown for contact import.
- Color not shown to see the progress of the import.
- Improper width - need to scroll down within the tiny box to see the minutes remaining.

<img width="497" alt="image" src="https://user-images.githubusercontent.com/5929648/30683131-cb2217fe-9ec9-11e7-8637-31fb41a736cc.png">

After
----------------------------------------
Progress bar shown for Contribution, membership and activity imports with proper width and color.

<img width="528" alt="image" src="https://user-images.githubusercontent.com/5929648/30683431-11d159e8-9ecb-11e7-8eeb-f8dfc9976cf2.png">

---

 * [CRM-21204: Show Import Progress Bar for Activity, Contribution and Membership imports.](https://issues.civicrm.org/jira/browse/CRM-21204)

jitendrapurohit commented:
test this please

seamuslee001 commented:
Jenkins re test this please

jitendrapurohit commented:
jenkins, retest this please

colemanw commented on CRM/Import/Parser.php:
```diff
@@ -293,6 +293,58 @@ public function &getActiveFieldParams() {
   }
 
   /**
```
Can you please say what this function does in this comment?

jitendrapurohit replied:
Done.

colemanw commented on templates/CRM/common/importProgress.tpl:
```diff
@@ -0,0 +1,81 @@
+
+  function pollLoop() {
+    setIntermediate();
+    window.setTimeout(pollLoop, 10*1000); // 10 sec
```
This looks like a potential memory leak as the function is continually calling itself. Also I don't see how this ever ends. It should stop once it reaches 100% but I don't see where that would happen.

colemanw commented on templates/CRM/common/importProgress.tpl:
```diff
@@ -0,0 +1,81 @@
+    });
+    $("#importProgressBar" ).progressbar({value:0});
+    $("#importProgressBar").show( );
+    pollLoop( );
```
I think it would be better to do

    loop = window.setInterval(setIntermediate, 5)

here and delete the function `pollLoop`, and declare the `var loop` as the first line of this closure (line 29 just after the `CRM.$(function(){`) that way it is accessible from all functions and can be cleared by `setIntermediate` once progress is 100%.

jitendrapurohit replied:
@colemanw I've pushed the changes as per the above suggestion. Thanks 👍

jitendrapurohit commented:
test this please

mlutfy commented:
@jitendrapurohit This seems good, but is a fairly large patch. Has Fuzion deployed it to production? Has it had "real" (user) testing?

jitendrapurohit commented:
Thanks for the response @mlutfy. We have applied this patch to our Fuzion civi tarballs(civi release + fuzion patches) which is being applied to all our client sites. So yes, we have deployed this to production and we haven't received any issues for the same. I'll check with other members who have tested this patch and let you know.

mlutfy commented on templates/CRM/common/importProgress.tpl:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  function verify() {
+    if (!confirm('Backing up your database before importing is recommended, as there is no Undo for this. {/literal}{ts escape='js'}Are you sure you want to Import now{/ts}{literal}?')) {
```
Please fix the `{ts}` usage in this string:  the part "Backing up database before importing ..." should be in a `ts`.

Ex: `if (!confirm('{/literal}{ts escape='js'}Backuping up ....{/ts} {ts escape='js'}Are you sure ...?{/ts}{literal}'))`

Also: `{ts escape='js'}Are you sure you want to Import now{/ts}{literal}?'`  -> the interrogation should be inside the `{ts}`.

mlutfy commented:
I applied the PR to a set of our production sites and did a bit of testing, seems to work well. If you don't mind fixing the "ts" string (I ignored the other ones, which also pre-date this patch, less critical), considering that Coleman already reviewed (and subsequent improvements), I think this will be good to merge.

colemanw commented on templates/CRM/common/importProgress.tpl:
```diff
@@ -0,0 +1,77 @@
+      if (result[0] < 100) {
+        $("#importProgressBar .ui-progressbar-value").animate({width: result[0] + "%"}, 500);
+        $("#status").text(result[0] + "% Completed");
+      }
```
else {
      window.clearInterval(loop);
    }

jitendrapurohit commented:
@mlutfy @colemanw Thanks for the comments. I've updated them all. I've also added a fix for an additional issue noticed on further testing, i.e, the `Cancel` button on the alert pop up used to continue the import instead of remaining on the same screen.
